yoyodyne/models/expert.py

```diff
@@ -408,82 +431,64 @@ def get_expert(
-        Training is quicker if tensors are converted to lists.
-        For efficiency, we encode action vocabulary simultaneously.
+        We want just the encodings without BOS or EOS tokens. This
+        encodes just raw source-target text for the Maxwell library.
 
         Args:
             data (data.Dataset): Dataset to iterate over.
-            actions (ActionVocabulary): Vocabulary object
-                to encode actions for expert.
+
         Returns:
             Iterator[Tuple[List[int], List[int]]]: Iterator that
                 yields list version of source and target entries
                 in dataset.
         """
-        for item in data:
-            # Dataset encodes BOW and EOW symbols for source. EOW
-            # for target. Removes these for SED training.
-            source = item.source.tolist()[1:-1]
-            target = item.target.tolist()[:-1]
-            actions.encode_actions(target)
-            if data.index.tie_embeddings:
-                actions.encode_actions(source)
-            yield source, target
-
-    def _encode_action_vocabulary(
-        data: data.Dataset, actions: ActionVocabulary
-    ) -> None:
-        """Encodes action vocabulary for expert oracle.
-
-        For instantiating SED objects from file.
-
-        Args:
-            data (data.Dataset): Dataset to iterate over.
-            actions (ActionVocabulary): Vocabulary object
-                to encode actions for expert.
-        """
-        for item in data:
-            # Ignores last symbol since EOW.
-            target = item.target.tolist()[:-1]
-            actions.encode_actions(target)
-            if data.index.tie_embeddings:
-                source = item.source.tolist()[1:-1]
-                actions.encode_actions(source)
-
-    actions = ActionVocabulary(unk_idx=train_data.index.unk_idx)
+        assert data.has_target, """Passed dataset with no target to expert
+            module, cannot perform SED!"""
+        for sampl in data.samples:
+            source, target = sampl[0], sampl[-1]
+            yield [data.index(s) for s in source], [
+                data.index(t) for t in target
+            ]
+
+    actions = ActionVocabulary(train_data.index)
     if sed_params_path:
-        sed_params = sed.ParamDict.read_params(sed_params_path)
-        sed_aligner = sed.StochasticEditDistance(sed_params)
-        # Loads vocabulary into action vocabulary.
-        _encode_action_vocabulary(train_data, actions)
-    else:
-        sed_aligner = sed.StochasticEditDistance.fit_from_data(
-            _generate_data_and_encode_vocabulary(train_data, actions),
-            epochs=epochs,
+        if epochs:
+            sed_aligner = sed.StochasticEditDistance.fit_from_data(
+                _generate_data(train_data),
+                epochs=epochs,
+            )
+            sed_aligner.params.write_params(sed_params_path)
+        else:
+            sed_params = sed.ParamDict.read_params(sed_params_path)
```

Comment:
Am I understanding correctly that in order to use an existing sed, the user would specify `0` or `None` for `oracle_em_epochs`? Avoiding rerunning em every time is a great feature, but I wonder if we can think through a more intuitive user interface for it, this feels a bit buried. Maybe you've already put some thought into this though so please let me know if you think this is already a good interface.

Reply:
Oh sorry I see the comment in the method header. I missed it before.

Reply:
I just added a bool to the train script that's triggered if a sed file already exists. I think this is a bit cleaner. Thoughts?

Reply:
Yeah that works. I am having trouble wrapping my head back around all of this now, so I think if we want further updates to this interface those can go in a later issue after we use the transducer more (which I need to do for a yoyodyne resubmission anyway).